### DISCUSSION TITLE
제공되는 코드를 보고, 자바스크립트에서 객체의 동등성은 어떻게 비교되는 지 설명해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#62] [원시 값과 객체의 비교]
- 제공되는 코드를 보고, 자바스크립트에서 객체의 동등성은 어떻게 비교되는 지 설명해주세요.

```js
var person1 = {
  name: 'Lee'
};

var person2 = {
  name: 'Lee'
};

console.log(person1 === person2);
console.log(person1.name === person2.name);
```

```js
var person1 = {
  name: 'Lee'
};

var person2 = person1;

console.log(person1 === person2);
console.log(person1.name === person2.name);
```

  - 두 객체가 동일하다고 판단하기 위해 어떤 방법을 사용할 수 있나요?
    - === 연산자와 Object.is 메서드의 차이는 무엇인가요?

**Comment by pipisebastian** [comment]
> 제공되는 코드를 보고, 자바스크립트에서 객체의 동등성은 어떻게 비교되는 지 설명해주세요.

- 1번째 코드의 경우, false, true가 반환됩니다. person1, person2는 서로 다른 메모리주소를 참조하기 때문에, 두객체는 같지 않지만, name 속성의 경우 같은 원시값이므로 true를 반환합니다.

- 2번째 코드의 경우, 모두 true를 반환합니다. 할당연산자로 받을 경우, 두 식별자는 모두 같은 메모리주소를 참조합니다. 이에 모두 true를 반환합니다.

> 두 객체가 동일하다고 판단하기 위해 어떤 방법을 사용할 수 있나요?
- `===` 일치연산자와, `Object.is` 를 사용할 수 있습니다. 

> === 연산자와 Object.is 메서드의 차이는 무엇인가요?
- `===` : NaN === NaN은 false이고, +0 === -0은 true
- `Object.is` :  NaN === NaN은 true를, +0 === -0은 false

**Comment by publdaze** [comment]
> - 제공되는 코드를 보고, 자바스크립트에서 객체의 동등성은 어떻게 비교되는 지 설명해주세요.
> 
> ```js
> var person1 = {
>   name: 'Lee'
> };
> 
> var person2 = {
>   name: 'Lee'
> };
> 
> console.log(person1 === person2);
> console.log(person1.name === person2.name);
> ```
> 
> ```js
> var person1 = {
>   name: 'Lee'
> };
> 
> var person2 = person1;
> 
> console.log(person1 === person2);
> console.log(person1.name === person2.name);
> ```
- 첫번째 코드를 보면 각각의 변수에 동일한 프로퍼티를 가지는 객체가 할당되어 있다.
- 객체 프로퍼티 값 자체는 같지만 프로퍼티가 저장된 메모리 주소가 다르기 때문에 첫번째 출력은 false를 반환하고,
- 두번째 출력은 프로퍼티 값 자체가 원시값이기 때문에 값 자체가 비교되어 true가 반환된다.
- 두번째 코드를 보면 person2에 person1이 복사가 됨.
- 참조값이 복사가 된 것이므로 두 변수 모두 동일한 참조 값을 가지기 때문에 첫번째 출력은 true가 반환됨
- 두번째 출력은 첫번째 코드와 마찬가지로 true가 반환됨

>   - 두 객체가 동일하다고 판단하기 위해 어떤 방법을 사용할 수 있나요?
- Object.is 메소드를 사용하거나, 동등 연산자 혹은 일치 연산자를 사용할 수 있음
>   - === 연산자와 Object.is 메서드의 차이는 무엇인가요?
- 동등연산자와 일치 연산자의 경우 +0과 -0은 동일하다고 평가하고, NaN과 NaN을 비교하면 다른 값으로 평가한다. Object.is 메서드는 비교적 정확하게 +0과 -0는 false, NaN과 NaN은 true로 평가한다.
